virttest/cartesian_config.py

```diff
@@ -538,10 +538,64 @@ def dump(self, indent, recurse=False):
 match_substitute = re.compile("\$\{(.+?)\}")
 
 
+def _drop_suffixes(d):
+    """
+    Merge suffixes for same var, or drop off unnecessary suffixes
+
+    This step returns a copy of a suffix flattened dictionary.
+    """
+    # dictionary `d_flat' is going to be the mutated copy of `d`
+    d_flat = d.copy()
+    for key in d:
+        if key in _reserved_keys:
+            continue
+
+        if not isinstance(key, tuple):
+            continue
+
+        try:
+            # This file was invoked through cmdline
+            options.skipdups
+            skipdups = options.skipdups
+        except NameError:
+            # This file was invoked as Python module
+            skipdups = True
+
+        if skipdups:
+            # Drop vars with suffixes matches general var val
+            # Example: if a_x == 1, and a == 1. Drop: a_x, leave a
```

Comment:
@pevogam 
 Yet to completely understand the usecase of this, here is what I understood, correct me if wrong:
patch has two parts:
1. flaten the keys 
        Ex:
```
        newvcpucores = 12
         vms="vm1 vm2 vm3 vm4"
         cores = 4
         cores_vm1 = 4
         cores_vm2 = 8
         test1:
              cores= ${newvcpucores}
```
        without this patch:
                    so core values of vm1 will 4, vm2 will be 8 and vm3,vm4 will be 12
         with this patch:
                   cores_vm1 will be dropped so 
                   so core values of vm1 will also be 12???, vm2 will be 8 and vm3,vm4 will be 12, but that's not what we want right?
          
           Or am I totally getting it wrong?

2. then attempt substitution [earlier it would directly attempt to substitute]

somehow commit message is  not explain to me

Reply:
Yes, the main idea is that substitution won't work on suffixed/joined variants which need flattening before being able to substitute.  Otherwise no substitution is possible for users using these variants like us.

>     without this patch:
>                 so core values of vm1 will 4, vm2 will be 8 and vm3,vm4 will be 12
>      with this patch:
>                cores_vm1 will be dropped so 
>                so core values of vm1 will also be 12???, vm2 will be 8 and vm3,vm4 will be 12, but that's not what we want right?
>       
>        Or am I totally getting it wrong?

Not sure what you mean here, if I run your example on the current branch the output will be:
```
> cat t1.cfg
newvcpucores = 12
vms="vm1 vm2 vm3 vm4"
cores = 4
cores_vm1 = 4
cores_vm2 = 8
test1:
     cores= ${newvcpucores}
> python3 virttest/cartesian_config.py -c t1.cfg
dict    1:  
    cores = 4
    cores_vm1 = 4
    cores_vm2 = 8
    dep = []
    name = 
    newvcpucores = 12
    shortname = 
    vms = vm1 vm2 vm3 vm4
```
If I run it slightly corrected:
```
> cat t1.cfg
newvcpucores = 12
vms="vm1 vm2 vm3 vm4"
cores = 4
cores_vm1 = 4
cores_vm2 = 8
test1:
     cores= ${newvcpucores}
> python3 virttest/cartesian_config.py -c t1.cfg
dict    1:  
    cores = 4
    cores_vm1 = 4
    cores_vm2 = 8
    dep = []
    name = 
    newvcpucores = 12
    shortname = 
    vms = vm1 vm2 vm3 vm4

> cat t1.cfg 
newvcpucores = 12
vms="vm1 vm2 vm3 vm4"
cores = 4
cores_vm1 = 4
cores_vm2 = 8
#test1:
cores= ${newvcpucores}
> python3 virttest/cartesian_config.py -c t1.cfg
dict    1:  
    cores = 12
    cores_vm1 = 4
    cores_vm2 = 8
    dep = []
    name = 
    newvcpucores = 12
    shortname = 
    vms = vm1 vm2 vm3 vm4    
```
I obtain your "without this patch" version but with the patch.

Reply:
Regarding the commit message, this is exactly what I wrote with
```
Joint variants need flattening and dropping of indices in order
to detect values that should be substituted if a previous parameter
of the given key exists.
```
but if I miss something let me know and I will correct it.

Reply:
Ping @sathnaga, sorry if I bother you.